qualtran/bloqs/chemistry/thc_notebook_utils.py

```diff
@@ -0,0 +1,147 @@
+    """Genereralizer for THC prepare bloqs."""
+    if isinstance(bloq, (Allocate, Free, Split, Join)):
+        return None
+    if isinstance(bloq, (Rx, Ry, Rz)):
+        return attrs.evolve(bloq, angle=phi_sym)
+    if isinstance(bloq, CirqGateAsBloq):
+        if isinstance(bloq.gate, cirq_ft.algos.And) and (len(bloq.gate.cv) == 2):
+            return And(cv1=and_cv0, cv2=and_cv1, adjoint=bloq.gate.adjoint)
+        if isinstance(bloq.gate, cirq_ft.algos.SwapWithZeroGate):
+            return SwapWithZero(
+                bloq.gate.selection_bitsize, bloq.gate.target_bitsize, bloq.gate.n_target_registers
+            )
+        # How to check type of cirg_gate**-1
+        if (
+            isinstance(bloq.gate, cirq.ops.raw_types._InverseCompositeGate)
+            and 'SwapWithZero' in bloq.pretty_name()
+        ):
+            return SwapWithZero(
+                (bloq.gate**-1).selection_bitsize,
+                (bloq.gate**-1).target_bitsize,
+                (bloq.gate**-1).n_target_registers,
+            )
+        if isinstance(bloq.gate, cirq_ft.algos.SelectSwapQROM):
+            bloq.gate.__class__.__repr__ = custom_qroam_repr
+        if isinstance(bloq.gate, cirq_ft.algos.QROM):
+            bloq.gate.__class__.__repr__ = custom_qrom_repr
+        if isinstance(bloq.gate, single_qubit_clifford):
+            return ArbitraryClifford(n=1)
+        if isinstance(bloq.gate, two_qubit_clifford):
+            return ArbitraryClifford(n=2)
+        if (
+            isinstance(bloq.gate, cirq.ControlledGate)
+            and isinstance(bloq.gate._sub_gate, single_qubit_clifford)
+            and bloq.gate.num_controls() == 1
+        ):
+            return ArbitraryClifford(n=2)
+    return bloq
+
+
+def bin_bloq_counts(bloq):
+    """Classify bloq counts.
+
+    It's helpful to classify bloqs by their type (comparators, reflections, swaps, ...)
+
+    TODO: Generalize to take in a dict and custom classifier callable similar to
+    generalize for bloq counts.
+
+    Returns
+    -------
+    classified_bloqs : Dictionary of type of bloq and the sum of the counts of
+        bloqs of this type.
+    """
+    tot_t = 0
+    classified_bloqs = defaultdict(list)
+    for num_calls, bloq in bloq.bloq_counts():
+        if isinstance(bloq, (Split, Join, Allocate, Free)):
+            pass
+        else:
+            num_t = get_bloq_counts_graph(bloq, generalizer=generalize)[1].get(TGate())
+            if num_t is not None:
+                tot_t += num_calls * num_t
+                if isinstance(bloq, bloq_comparators):
+                    classified_bloqs['comparator'] += [num_calls * num_t]
+                elif isinstance(bloq, CirqGateAsBloq):
+                    if isinstance(bloq.gate, cirq_ft.MultiControlPauli) and isinstance(
+                        bloq.gate.target_gate, cirq.ops.common_gates.ZPowGate
+                    ):
+                        classified_bloqs['reflections'] += [num_calls * num_t]
```

Comment:
why are you extending a list for each one?

Did you mean to do

```python
d = defaultdict(lambda: 0)

for whatever:
  d[key] += 5
```

Reply:
I wanted to keep a list as (at least here) there are multiple comparators of different size so I could eyeball which ones are different that what's quoted in the paper vs what's implemented. For example, maybe the comparison to a constant was different but the two register comparison was correct. It's sort of helpful for debugging.

Reply:
Well at least it was originally this way, I see I am summing everything together at the end.

Reply:
Even if they were kept as a list, there's no way to know which entry corresponded to which bloq, really

Reply:
In general no, In this specific case I knew one was larger than the other (and there were only two or three possibilities). I've converted to a dataclass of ints though, as it's no longer relevent.